src/OperatorEnumConstruction.jl

```diff
@@ -51,108 +52,159 @@ function create_evaluation_helpers!(operators::GenericOperatorEnum)
             end
             return out
         end
+        function Base.adjoint(::Node{T}) where {T}
+            return _ -> begin
+                error("Gradients are not implemented for `GenericOperatorEnum`.")
+            end
+        end
     end
 end
 
-function create_construction_helpers!(
-    operators::AbstractOperatorEnum; extend_user_operators::Bool=false
-)
-    is_scalar_operator_enum = typeof(operators) <: OperatorEnum
-    type_requirements = is_scalar_operator_enum ? Real : Any
-
-    for (op, f) in enumerate(map(Symbol, operators.binops))
-        if is_scalar_operator_enum
-            f = if f in [:pow, :safe_pow]
-                Symbol(^)
-            else
-                f
+function _extend_unary_operator(f::Symbol, op, type_requirements)
+    quote
+        quote
+            function $($f)(l::Node{T})::Node{T} where {T<:$($type_requirements)}
+                return if (l.degree == 0 && l.constant)
+                    Node(T; val=$($f)(l.val::T))
+                else
+                    Node($($op), l)
+                end
             end
         end
-        if isdefined(Base, f)
-            f = :(Base.$(f))
-        elseif !extend_user_operators
-            # Skip non-Base operators!
-            continue
-        end
-        Base.MainInclude.eval(
-            quote
-                import DynamicExpressions: Node
+    end
+end
 
-                function $f(l::Node{T}, r::Node{T}) where {T<:$type_requirements}
-                    if (l.degree == 0 && l.constant && r.degree == 0 && r.constant)
-                        Node(T; val=$f(l.val::T, r.val::T))
-                    else
-                        Node($op, l, r)
-                    end
+function _extend_binary_operator(f::Symbol, op, type_requirements, build_converters)
+    quote
+        quote
```

Comment:
This double quote is a bit cumbersome, but is required since we do not know the function names beforehand. The code:
```julia
for f in [:vec_add, :vec_plus]
    function f(x)
        x
    end
end
```
will not define `vec_add`, it will try to define a function `f`. Instead, we need to use:
```julia
for f in [:vec_add, :vec_plus]
    eval(quote
        function $(f)(x)
            x
        end
    end)
end
```
However, we want to generate this code inside a macro. So, the double quote is necessary to build the inner part of this snippet.

Reply:
I'm not sure I understand the double quote and `$($f)` business. You're also calling `eval` later on, which is a hint that things could be refactored. 

```Julia
julia> module A
           struct MyType end
           macro foo(f)
               return quote
                   function $(esc(f))(x::MyType)
                       return x
                   end
               end
           end
       end
Main.A

julia> module B
           using Main.A
           bar() = 1
           A.@foo bar
       end
Main.B

julia> B.bar
bar (generic function with 2 methods)

julia> methods(B.bar)
# 2 methods for generic function "bar":
[1] bar() in Main.B at REPL[2]:3
[2] bar(x::Main.A.MyType) in Main.B at REPL[1]:5
```

Reply:
Thanks! Although, would this work for passing a collection of functions to the macro? Since the functions will be defined inside a loop:
```
for i=1:2
   function f(x)
       return i
   end
end
# f not defined
```
This is the reason I am doing the `eval` inside the macro. Not sure if there's an alternative?

Reply:
Here's a simpler example of what the double quote business is doing:
```julia
struct MyType end
for op in [:+, :-]
    @eval function Base.$(op)(x::MyType, y::MyType)
        return x
    end
end
hasmethod((+), (MyType, MyType))  # true
```
The macro is essentially returning code like this, to extend all user operators to the `Node` type.